File: java/src/main/java/shapes/client/shell/Runner.java
package shapes.client.shell;

import static shapes.client.shell.InputReader.readInput;

import java.util.Map;

public final class Runner {
    public Runner() {
        System.out.println("=".repeat(53));
        System.out.println("Welcome to the Area and Perimeter Calculator!");
        System.out.println("=".repeat(53));
    }

    static void startingMessage() {
        System.out.println("Choose a shape to calculate its area and perimeter.");
        System.out.println("  [C] CIRCLE");
        System.out.println("  [R] RECTANGLE");
        System.out.println("  [S] SQUARE");
        System.out.println("Type '[E]XIT' to quit the program.");
    }

    public void run() {
        Runnable printThankYouMessage = () -> {
            System.out.println("Thank you for using the Shapes CLI!");
            System.out.println("P.S. Did you know? The secret code is 42!");
        };

        while (true) {
            startingMessage();

            String shape = readInput();
            if (shape.equals("q")) {
                printThankYouMessage.run();
                break;
            }

            Map<String, Runnable> actions = Map.ofEntries(
                Map.entry("circle", Handler::handleCircle),
                Map.entry("c", Handler::handleCircle),
                Map.entry("rectangle", Handler::handleRectangle),
                Map.entry("r", Handler::handleRectangle),
                Map.entry("square", Handler::handleSquare),
                Map.entry("s", Handler::handleSquare)
            );

            var action = actions.getOrDefault(shape, () -> {
                System.out.println("Invalid input. Please try again.");
                System.out.println("Type one of the listed options.");
            });

            action.run();

            System.out.println("=".repeat(53));
        }
    }
}

File: csharp/Src/Shapes/Client/Shell/Runner.cs
using static Shapes.Client.Shell.InputReader;

namespace Shapes.Client.Shell;

public sealed class Runner {
    public Runner() {
        Console.WriteLine(new string('=', 53));
        Console.WriteLine("Welcome to the Area and Perimeter Calculator!");
        Console.WriteLine(new string('=', 53));
    }

    private static void StartingMessage() {
        Console.WriteLine(
            "Choose a shape to calculate its area and perimeter."
        );
        Console.WriteLine("  [C] CIRCLE");
        Console.WriteLine("  [R] RECTANGLE");
        Console.WriteLine("  [S] SQUARE");
        Console.WriteLine("Type '[E]XIT' to quit the program.");
    }

    public void Run() {
        Action printThankYouMessage = () => {
            Console.WriteLine("Thank you for using the Shapes CLI!");
            Console.WriteLine("P.S. Did you know? The secret code is 42!");
        };

        while (true) {
            Runner.StartingMessage();

            string shape = ReadInput();
            if (shape == "q") {
                printThankYouMessage();
                break;
            }

            Dictionary<string, Action>? actions = new() {
                ["circle"] = Handler.HandleCircle,
                ["c"] = Handler.HandleCircle,
                ["rectangle"] = Handler.HandleRectangle,
                ["r"] = Handler.HandleRectangle,
                ["square"] = Handler.HandleSquare,
                ["s"] = Handler.HandleSquare,
            };

            var action = actions.TryGetValue(shape, out Action? value)
                ? value
                : () => {
                    Console.WriteLine("Invalid input. Please try again.");
                    Console.WriteLine("Type one of the listed options.");
                };

            action();

            Console.WriteLine(new string('=', 53));
        }
    }
}
